fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -959,3 +931,188 @@ Future<Standalone<RangeResultRef>> ExclusionInProgressRangeImpl::getRange(ReadYo
                                                                           KeyRangeRef kr) const {
 	return ExclusionInProgressActor(ryw, getKeyRange().begin, kr);
 }
+
+Standalone<RangeResultRef> rywModuleGetRange(ReadYourWritesTransaction* ryw, Standalone<RangeResultRef> res,
+                                             KeyRangeRef kr) {
+	// res is read from database, if ryw enabled, we update it with writes in the transaction
+	Standalone<RangeResultRef> result;
+	RangeMap<Key, std::pair<bool, Optional<Value>>, KeyRangeRef>::Ranges ranges =
+	    ryw->getSpecialKeySpaceWriteMap().containedRanges(kr);
+	RangeMap<Key, std::pair<bool, Optional<Value>>, KeyRangeRef>::iterator iter = ranges.begin();
+	int index = 0;
+	while (iter != ranges.end()) {
```

Comment:
If I'm following, this should take the union of two sorted lists of key-value pairs, preferring the kv pair from the getSpecialKeySpaceWriteMap if they both have the same key. Can we structure it more like the usual algorithm for that?
```
while (iter1 != end1 || iter1 != end2) {
  if (iter1 == end1) {
    ++iter2;
  } else if (iter2 == end2) {
    ++iter1;
  } else if (iter1->key < iter2->key) {
    ++iter1;
  } else if (iter1->key == iter2->key) {
    ++iter1;
    ++iter2;
  } else if (iter1->key > iter2->key) {
    ++iter2;
  }
}
```

Reply:
Refactor code in this way. Quite an elegant way!
Thanks.